### PR TITLE
Fe/local storage

### PR DESCRIPTION
**Pull request by pluz-stas**

### CODE REVIEW

File: SimpleChat.Client/Shared/UserModel.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        {
+            UserData = new UserViewModel();
+        }
+        public UserViewModel UserData { get; set; }
+        protected async Task SetAsync()
+        {
+            await LocalStorageService.SetAsync(UserStateKeyName, UserData);
+        }
+    }
+    public class UserViewModel
```
Comment (chelovek1418) [inline SimpleChat.Client/Shared/UserModel.cs]: Do not store several classes in one file

File: SimpleChat.Client/Shared/UserModel.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+using SimpleChat.Client.Infrastructure;
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace SimpleChat.Client.Shared
+{
+    public class UserModel : ComponentBase
```
Comment (chelovek1418) [inline SimpleChat.Client/Shared/UserModel.cs]: Why do you inherit ComponentBase?

File: SimpleChat.Client/Shared/UserModel.cs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    {
+        [Inject] public ILocalStorageService LocalStorageService { get; set; }
+        internal const string UserStateKeyName = "UserState";
+        public UserModel() 
+        {
+            UserData = new UserViewModel();
+        }
+        public UserViewModel UserData { get; set; }
+        protected async Task SetAsync()
```
Comment (chelovek1418) [inline SimpleChat.Client/Shared/UserModel.cs]: Protected? Consider using the strictest access modifiers

File: SimpleChat.Client/Shared/UserModel.cs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace SimpleChat.Client.Shared
+{
+    public class UserModel : ComponentBase
+    {
+        [Inject] public ILocalStorageService LocalStorageService { get; set; }
```
Comment (chelovek1418) [inline SimpleChat.Client/Shared/UserModel.cs]: Do you really need public modifier here? Consider using the strictest access modifiers

File: SimpleChat.Client/Shared/UserComponent.razor [hunk 5]
```diff
@@ -1 +1,17 @@
-﻿<h3 @attributes="@InputAttributes">UserComponent</h3>
\ No newline at end of file
+﻿@inherits UserModel
```
Comment (chelovek1418) [inline SimpleChat.Client/Shared/UserComponent.razor]: Do not use this approach with inheritance. Consider using partial classes instead

File: SimpleChat.Client/Infrastructure/LocalStorageService.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+        public LocalStorageService(IJSRuntime jsRuntime)
+        {
+            _jsRuntime = jsRuntime;
+        }
+
+        public async Task SetAsync<T>(string key, T item) where T : class
+        {
+            var data = JsonSerializer.Serialize(item);
+            await _jsRuntime.InvokeVoidAsync("set", key, data);
```
Comment (chelovek1418) [inline SimpleChat.Client/Infrastructure/LocalStorageService.cs]: Move hardcode strings to constants

File: SimpleChat.Client/Infrastructure/LocalStorageService.cs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+            _jsRuntime.InvokeAsync<string>("set", key, value);
+            return Task.CompletedTask;
+        }
+
+        public async Task<T> GetAsync<T>(string key) where T : class
+        {
+            var data = await _jsRuntime.InvokeAsync<string>("get", key);
+            if (string.IsNullOrEmpty(data))
+            {
+                return null!;
```
Comment (chelovek1418) [inline SimpleChat.Client/Infrastructure/LocalStorageService.cs]: Exclamation mark?

File: SimpleChat.Client/wwwroot/js-local-storage.js [hunk 8]
```diff
@@ -0,0 +1,4 @@
+﻿function message(e) { alert(e); }
```
Comment (chelovek1418) [inline SimpleChat.Client/wwwroot/js-local-storage.js]: Use typeScript instead

File: SimpleChat.Client/Infrastructure/ILocalStorageService.cs [hunk 9]
```diff
@@ -0,0 +1,17 @@
+﻿using System.Threading.Tasks;
+
+namespace SimpleChat.Client.Infrastructure
+{
+    public interface ILocalStorageService
+    {
+        Task SetAsync<T>(string key, T item) where T : class;
```
Comment (chelovek1418) [inline SimpleChat.Client/Infrastructure/ILocalStorageService.cs]: Drop this method

File: SimpleChat.Client/Infrastructure/ILocalStorageService.cs [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+        Task SetStringAsync(string key, string value);
+
+        Task<T> GetAsync<T>(string key) where T : class;
```
Comment (chelovek1418) [inline SimpleChat.Client/Infrastructure/ILocalStorageService.cs]: And this

File: SimpleChat.Client/Models/UserModel.cs [hunk 11]
```diff
@@ -0,0 +1,13 @@
+﻿using System.ComponentModel.DataAnnotations;
```
Comment (chelovek1418) [inline SimpleChat.Client/Models/UserModel.cs]: Drop redundant model

File: SimpleChat.Client/Shared/UserComponent.razor.cs [hunk 12]
```diff
@@ -1,11 +1,79 @@
...
+using SimpleChat.Client.Infrastructure;
+using SimpleChat.Client.Models;
 
 namespace SimpleChat.Client.Shared
 {
     public partial class UserComponent
     {
         [Parameter(CaptureUnmatchedValues = true)]
         public Dictionary<string, object> InputAttributes { get; set; }
+        [Inject] private ILocalStorageService LocalStorageService { get; set; }
```
Comment (chelovek1418) [inline SimpleChat.Client/Shared/UserComponent.razor.cs]: Separate for two lines

File: SimpleChat.Client/Shared/UserComponent.razor.cs [hunk 13]
```diff
@@ -1,11 +1,79 @@
...
+
+        [Required]
+        [StringLength(50, ErrorMessage = "Too Long")]
+        private string UserName { get; set; }
+        private string UserImg { get; set; }
+
+        protected override async Task OnInitializedAsync()
+        {
+            string localStorageName = await LocalStorageService.GetStringAsync(UserNameKeyName);
+            if (localStorageName == null)
```
Comment (chelovek1418) [inline SimpleChat.Client/Shared/UserComponent.razor.cs]: Consider using ternary operator

Comment (chelovek1418) [reply SimpleChat.Client/Shared/UserComponent.razor.cs]: And use string method IsNullOrWhiteSpace for checking string

File: SimpleChat.Client/Shared/UserComponent.razor.cs [hunk 14]
```diff
@@ -1,11 +1,79 @@
...
+        {
+            if (UserImg != null)
+            {
+                await LocalStorageService.SetStringAsync(UserImgKeyName, UserName);
+            }
+        }
+        
+        private async Task ToggleEditUserNameInputAsync()
+        {
+            if (!isUserImgInputOpen)
```
Comment (chelovek1418) [inline SimpleChat.Client/Shared/UserComponent.razor.cs]: attention!!

File: SimpleChat.Client/wwwroot/js-local-storage.ts [hunk 15]
```diff
@@ -0,0 +1,3 @@
+﻿function set(key, value) { localStorage.setItem(key, value); }
```
Comment (chelovek1418) [inline SimpleChat.Client/wwwroot/js-local-storage.ts]: Add data type for params

File: SimpleChat.Client/Shared/UserComponent.razor.cs [hunk 16]
```diff
@@ -1,11 +1,72 @@
...
+            UserName = IsNullOrWhiteSpace(localStorageName) ? "anon" : localStorageName;
+            UserImg = await LocalStorageService.GetStringAsync(UserImgKeyName);
+            await base.OnInitializedAsync();
+        }
+        
+        private async Task SetNameAsync()
+        {
+            if (!IsNullOrWhiteSpace(UserName))
+            {
+                Console.WriteLine("40");
```
Comment (chelovek1418) [inline SimpleChat.Client/Shared/UserComponent.razor.cs]: What is it?

Comment (pluz-stas) [reply SimpleChat.Client/Shared/UserComponent.razor.cs]: ska

**Pull request merged**